### PR TITLE
feature(aws-s3-connector): AWS s3 connector

### PR DESCRIPTION
**Pull request by mathias-vandaele**
## Related issues

closes #

## Checklist

- [ ] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+public record DeleteS3Action(
+    @TemplateProperty(
+            label = "AWS bucket",
+            id = "deleteActionBucket",
+            group = "deleteObject",
+            tooltip = "Bucket from where an object should be deleted",
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.bucket"))
+        @Valid
+        @NotNull
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java:25]: NotBlank ?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@TemplateSubType(id = "deleteObject", label = "Delete document")
+public record DeleteS3Action(
+    @TemplateProperty(
+            label = "AWS bucket",
+            id = "deleteActionBucket",
+            group = "deleteObject",
+            tooltip = "Bucket from where an object should be deleted",
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.bucket"))
+        @Valid
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java:24]: Not needed I guess

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        @NotNull
+        String bucket,
+    @TemplateProperty(
+            label = "AWS key",
+            id = "deleteActionKey",
+            group = "deleteObject",
+            tooltip = "Key of the object which should be deleted",
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.key"))
+        @Valid
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DeleteS3Action.java:34]: same

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DownloadS3Action.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.key"))
+        @Valid
+        @NotNull
+        String key,
+    @TemplateProperty(
+            label = "Create document",
+            id = "downloadActionAsFile",
+            group = "downloadObject",
+            tooltip = "....",
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/DownloadS3Action.java:41]: todo ?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        @Valid
+        String key,
+    @TemplateProperty(
+            label = "Document",
+            group = "uploadObject",
+            id = "uploadActionDocument",
+            tooltip = "Document to be uploaded on AWS S3",
+            type = TemplateProperty.PropertyType.String,
+            feel = Property.FeelMode.required,
+            binding = @TemplateProperty.PropertyBinding(name = "action.document"))
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java:46]: why do we need to specify the binding?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  public Object execute(S3Action s3Action) {
+    return switch (s3Action) {
+      case DeleteS3Action deleteS3Action -> delete(deleteS3Action);
+      case DownloadS3Action downloadS3Action -> download(downloadS3Action);
+      case UploadS3Action uploadS3Action -> upload(uploadS3Action);
+    };
+  }
+
+  private Object upload(UploadS3Action uploadS3Action) {
+    Long contentLength = uploadS3Action.document().metadata().getSize();
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:63]: can metadata() be null? I don't think so but I wanted to highlight it anyway :)

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    @TemplateProperty(
+            label = "AWS key",
+            id = "uploadActionKey",
+            group = "uploadObject",
+            tooltip =
+                "Key of the uploaded object, if not given. The file name from the document metadata will be used",
+            optional = true,
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.key"))
+        @Valid
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java:37]: We can remove it

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    @TemplateProperty(
+            label = "AWS bucket",
+            id = "uploadActionBucket",
+            group = "uploadObject",
+            tooltip = "Bucket from where an object should be uploaded",
+            feel = Property.FeelMode.optional,
+            binding = @TemplateProperty.PropertyBinding(name = "action.bucket"))
+        @Valid
+        @NotNull
+        String bucket,
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java:27]: @NotBlank

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        String key,
+    @TemplateProperty(
+            label = "Document",
+            group = "uploadObject",
+            id = "uploadActionDocument",
+            tooltip = "Document to be uploaded on AWS S3",
+            type = TemplateProperty.PropertyType.String,
+            feel = Property.FeelMode.required,
+            binding = @TemplateProperty.PropertyBinding(name = "action.document"))
+        Document document)
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/UploadS3Action.java:47]: @NotNull?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            .key(
+                Optional.ofNullable(uploadS3Action.key())
+                    .orElse(uploadS3Action.document().metadata().getFileName()))
+            .contentLength(contentLength)
+            .contentType(contentType)
+            .build();
+
+    this.s3Client.putObject(
+        putObjectRequest,
+        RequestBody.fromInputStream(uploadS3Action.document().asInputStream(), contentLength));
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:78]: we could try with resource to close the stream?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+import software.amazon.awssdk.core.ResponseInputStream;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+public class S3Executor {
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:33]: maybe can we move the `response` package in the `model` package, not sure how we do usually

File: connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+class S3ConnectorFunctionTest extends BaseTest {
+
+  @ParameterizedTest
+  @MethodSource("loadUploadActionVariables")
+  void executeUploadActionReturnsCorrectResult(String variables) {
+
+    var bedrockConnectorFunction = new S3ConnectorFunction();
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java:30]: should rename this :p

File: connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      Assertions.assertNotNull(response);
+      Assertions.assertInstanceOf(UploadResponse.class, response);
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("loadDownloadActionVariables")
+  void executeDownloadActionReturnsCorrectResult(String variables) {
+
+    var bedrockConnectorFunction = new S3ConnectorFunction();
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java:48]: same

File: connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+      Assertions.assertNotNull(response);
+      Assertions.assertInstanceOf(DownloadResponse.class, response);
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("loadDeleteActionVariables")
+  void executeDeleteActionReturnsCorrectResult(String variables) {
+
+    var bedrockConnectorFunction = new S3ConnectorFunction();
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/S3ConnectorFunctionTest.java:66]: same

File: connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/BaseTest.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import io.camunda.connector.api.json.ConnectorsObjectMapperSupplier;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.stream.Stream;
+import org.junit.jupiter.params.provider.Arguments;
+
+public class BaseTest {
+
+  public static Stream<String> loadUploadActionVariables() {
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/BaseTest.java:23]: What about creating integration tests using https://java.testcontainers.org/modules/localstack/?

**Review by johnBgood — COMMENTED** [review]
Some questions, overall great!

File: connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/BaseTest.java (hunk 15 above)
Comment (mathias-vandaele) [reply connectors/aws/aws-s3/src/test/java/io/camunda/connector/aws/s3/BaseTest.java:23]: Yes, I might created a separated PR for integration tests.

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Environment:
  - [ ] SaaS only
  - [ ] SM only
  - [ ] Both
- Connectors version: <TO_BE_REPLACED>

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

File: connectors/aws/aws-s3/element-templates/aws-s3-outbound-connector.json [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+      "type" : "zeebe:input"
+    },
+    "type" : "Hidden"
+  }, {
+    "id" : "actionDiscriminator",
+    "label" : "Action",
+    "value" : "uploadObject",
+    "group" : "action",
+    "binding" : {
+      "name" : "actionDiscriminator",
```
Comment (sbuettner) [inline connectors/aws/aws-s3/element-templates/aws-s3-outbound-connector.json:141]: Can we just call it "action"?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/request/UploadS3Action.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+package io.camunda.connector.aws.s3.model.request;
+
+import io.camunda.connector.generator.dsl.Property;
+import io.camunda.connector.generator.java.annotation.TemplateProperty;
+import io.camunda.connector.generator.java.annotation.TemplateSubType;
+import io.camunda.document.Document;
+import jakarta.validation.constraints.NotBlank;
+
+@TemplateSubType(id = "uploadObject", label = "Upload document")
+public record UploadS3Action(
```
Comment (sbuettner) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/request/UploadS3Action.java:16]: We now have 3 names 😄 
id: `uploadObject`, 
label: `Upload document`
record: `UploadS3Action`

Should we go with a single one? Either `PutObject` or `UploadObject`?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/response/DownloadResponse.java [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. Licensed under a proprietary license.
+ * See the License.txt file for more information. You may not use this file
+ * except in compliance with the proprietary license.
+ */
+package io.camunda.connector.aws.s3.model.response;
+
+import io.camunda.document.Document;
+
+public record DownloadResponse(String bucket, String key, Document document, Object content) {}
```
Comment (sbuettner) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/model/response/DownloadResponse.java:11]: @mathias-vandaele Is it either a `Document` or the `content`? 
If so please create a simple sealed type which indicates that.

File: connectors/email/src/main/java/io/camunda/connector/email/client/jakarta/utils/JakartaUtils.java [hunk 19]
```diff
@@ -282,11 +284,19 @@ private void processBodyPart(
...
-              .getDisposition()
-              .equalsIgnoreCase(Part.ATTACHMENT) ->
+      case InputStream attachment when Part.ATTACHMENT.equalsIgnoreCase(
+              bodyPart.getDisposition()) ->
+          emailBodyBuilder.addAttachment(
+              new EmailAttachment(
+                  attachment,
+                  bodyPart.getFileName(),
+                  new ContentType(bodyPart.getContentType()).getBaseType()));
+      case String textAttachment when Part.ATTACHMENT.equalsIgnoreCase(bodyPart.getDisposition()) ->
```
Comment (sbuettner) [inline connectors/email/src/main/java/io/camunda/connector/email/client/jakarta/utils/JakartaUtils.java:294]: Is this change related to S3?

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+            .build();
+
+    this.s3Client.putObject(
+        putObjectRequest,
+        RequestBody.fromInputStream(uploadObject.document().asInputStream(), contentLength));
+
+    return new UploadResponse(
+        uploadObject.bucket(),
+        uploadObject.key(),
+        String.format("https://%s.s3.amazonaws.com/%s", uploadObject.bucket(), uploadObject.key()));
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:84]: we can prob introduce a constant here

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+                  .fileName(downloadObject.key())
+                  .build());
+    }
+  }
+
+  private DownloadResponse retrieveResponseWithContent(
+      String bucket, String key, ResponseInputStream<GetObjectResponse> responseResponseInputStream)
+      throws IOException {
+    byte[] rawBytes = responseResponseInputStream.readAllBytes();
+    return switch (responseResponseInputStream.response().contentType()) {
```
Comment (johnBgood) [inline connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:125]: not sure here, but be aware that sometimes the case is weird (uppercase letters etc), and also we might have `application/json; charset=utf-8`. Maybe it's not applicable here though.

Comment (johnBgood) [reply connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:125]: Maybe should we use [MimeType](https://docs.oracle.com/javase/8/docs/api///?javax/activation/MimeType.html) class for this

**Review by johnBgood — APPROVED** [review]
One question, but happy to approve :)

File: connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java (hunk 21 above)
Comment (mathias-vandaele) [reply connectors/aws/aws-s3/src/main/java/io/camunda/connector/aws/s3/core/S3Executor.java:125]: The ContentType returned should always be standard
image

**Comment by slolatte** [comment]
# QA information
## Test Environment
- Environment:
  - [ ] SaaS only
  - [ ] SM only
  - [ ] Both
- Connectors version: <TO_BE_REPLACED>

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Environment:
  - [ ] SaaS only
  - [ ] SM only
  - [ ] Both
- Connectors version: <TO_BE_REPLACED>

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*
